spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+    }
+  }
+
+  private List<ManifestFile> findMatchingManifests() {
+    Snapshot snapshot = table.currentSnapshot();
+    if (snapshot == null) {
+      return ImmutableList.of();
+    }
+    return snapshot.manifests().stream().filter(predicate).collect(Collectors.toList());
+  }
+
+  // computes the average manifest entry size based on available stats for manifests
+  private Long computeManifestEntrySizeBytes(List<ManifestFile> manifests) {
+    long totalSize = 0L;
+    int numEntries = 0;
+
+    for (ManifestFile m : manifests) {
+      ValidationException.check(
+          m.addedFilesCount() != null && m.existingFilesCount() != null && m.deletedFilesCount() != null,
+          "No file counts in manifest: " + m.path());
+
+      totalSize += m.length();
+      numEntries += m.addedFilesCount() + m.existingFilesCount() + m.deletedFilesCount();
+    }
+
+    ValidationException.check(totalSize > 0L, "Total size of manifests must be greater than 0");
+    ValidationException.check(numEntries > 0, "Number of manifest entries must be greater than 0");
+
+    return totalSize / numEntries;
+  }
+
+  // computes the estimated metadata size per spec and partition
+  private Map<Integer, List<PartitionMetadata>> computeMetadataSizeSummary(
+      Dataset<Entry> manifestEntryDS,
+      long manifestEntrySizeBytes) {
+
+    MetadataSizeAggregator agg = new MetadataSizeAggregator(specs, manifestEntrySizeBytes);
+    TypedColumn<Entry, Map<Integer, List<PartitionMetadata>>> column = agg.toColumn().name("result");
+    return manifestEntryDS.select(column).collectAsList().get(0);
+  }
+
+  // groups smaller partitions into bins of the target size
+  private Map<Integer, Map<StructLike, Integer>> computeBins(Map<Integer, List<PartitionMetadata>> sizeSummary) {
+    Map<Integer, Map<StructLike, Integer>> binMap = Maps.newHashMap();
+
+    sizeSummary.forEach((specId, sizes) -> {
+      BinPacking.ListPacker<PartitionMetadata> packer = new BinPacking.ListPacker<>(targetManifestSizeBytes, 1, false);
+      List<List<PartitionMetadata>> bins = packer.pack(sizes, PartitionMetadata::getMetadataSizeBytes);
+
+      for (int binIndex = 0; binIndex < bins.size(); binIndex++) {
+        List<PartitionMetadata> bin = bins.get(binIndex);
+
+        for (PartitionMetadata binEntry : bin) {
+          Map<StructLike, Integer> map = binMap.computeIfAbsent(binEntry.specId, key -> Maps.newHashMap());
+          map.put(binEntry.partition, binIndex);
+        }
+      }
+    });
+
+    return binMap;
+  }
+
+  private void replaceManifests(Iterable<ManifestFile> deletedManifests, Iterable<ManifestFile> addedManifests) {
+    try {
+      RewriteManifests rewriteManifests = table.rewriteManifests();
+      deletedManifests.forEach(rewriteManifests::deleteManifest);
+      addedManifests.forEach(rewriteManifests::addManifest);
+      summary.forEach(rewriteManifests::set);
+      rewriteManifests.commit();
+
+      if (!snapshotIdInheritanceEnabled) {
```

Comment:
This bit can be tricky if snapshot id inheritance is modified concurrently.

Reply:
I think we can pick up settings at the start of a commit and use the for the duration. There's no need to try to handle updates like that in the middle of a commit.

Reply:
I also don't want to overcomplicate this. The only problem is that the clean-up logic of staged manifests in this action depends on the value before the commit while the actual commit logic that decides whether to rewrite manifests will depend on calling `ops.current().properties()` in `BaseRewriteManifests`. If snapshot id inheritance is enabled concurrently, there can be a case when `BaseRewriteManifests` appends manifests without rewriting them but we still delete them after the commit. The probability of this is low as we need to reuse the same `Table` instance to enable snapshot id inheritance. It is just a point I wanted to bring attention to.

Reply:
I kept it simple for now.